src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/PostcodeSearch.cs

```diff
@@ -0,0 +1,194 @@
+                return CreateInvalidPostcodeResponse();
+            }
+
+            return vm;
+
+            ModelWithErrors<SearchCommand> CreateInvalidPostcodeResponse() =>
+                new ModelWithErrors<SearchCommand>(
+                    request,
+                    new ValidationResult(new[]
+                    {
+                        new ValidationFailure(nameof(SearchCommand.Postcode), "Enter a real postcode")
+                    }));
+        }
+
+        public async Task<OneOf<ModelWithErrors<SearchViewModel>, Success>> Handle(
+            SelectCommand request,
+            CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            var validator = new SelectCommandValidator();
+            var validationResult = validator.Validate(request);
+
+            if (!validationResult.IsValid)
+            {
+                // No address specified. Re-run the search with a prompt to select an address.
+                var vm = await CreateViewModel(request.Postcode);
+                return new ModelWithErrors<SearchViewModel>(vm, validationResult);
+            }
+
+            var addressDetail = await _addressSearchService.GetById(request.AddressId);
+
+            if (addressDetail == null)
+            {
+                // Address specified isn't valid. (We should never get here.)
+                // Re-run the search to get a list of valid addresses with a prompt to select an address.
+
+                var badAddressValidationResult = new ValidationResult(new[]
+                {
+                    new ValidationFailure(nameof(SelectCommand.AddressId), "Select an address")
+                });
+
+                var vm = await CreateViewModel(request.Postcode);
+                return new ModelWithErrors<SearchViewModel>(vm, badAddressValidationResult);
+            }
+
+            // SearchCommand has already checked that there is an ONSPD entry for this postcode
+            // so there is no valid case where `.Single()` should blow up
+            var onspdSearchResult = await _onspdSearchClient.Search(new OnspdSearchQuery() { Postcode = addressDetail.Postcode });
+            var onspdPostcodeRecord = onspdSearchResult.Items.Single();
+
+            _journeyInstance.UpdateState(state =>
+            {
+                state.AddressLine1 = addressDetail.Line1;
+                state.AddressLine2 = addressDetail.Line2;
+                state.Town = addressDetail.PostTown;
+                state.County = addressDetail.County;
+                state.Postcode = addressDetail.Postcode;
+                state.Latitude = onspdPostcodeRecord.Record.lat;
+                state.Longitude = onspdPostcodeRecord.Record.@long;
+                state.AddressIsOutsideOfEngland = onspdPostcodeRecord.Record.Country != "England";
+            });
+
+            return new Success();
+        }
+
+        private async Task<SearchViewModel> CreateViewModel(string postcode)
```

Comment:
This isn't "creating a view model", it's "performing a post code search". Consider extracting functionality or appropriate naming